apps/surfer/src/hooks/useFieldSelection.tsx

```diff
@@ -1,28 +1,91 @@
+import { ContentTypeId, ContentFieldsMap, ContentFieldId } from '../types';
+import { SurferCompatibility } from './useSurferCompatibility';
+
+export const useFieldSelection = (compatibility: SurferCompatibility = { compatibleFields: {}, compatibleContentTypes: [] }) => {
+  const [selectedContentTypes, setSelectedContentTypes] = useState<ContentTypeId[]>([]);
+  const [selectedContentFields, setSelectedContentFields] = useState<ContentFieldsMap>({});
+
+  const toggleContentTypeOff = (id: ContentTypeId) => {
+    setSelectedContentTypes((prevState) => prevState.filter((contentTypeId) => contentTypeId !== id));
+    setSelectedContentFields((prevState) => {
+      delete prevState[id];
+
+      return prevState;
+    });
+  };
+
+  const toggleContentTypeOn = useCallback(
+    (id: ContentTypeId, preSelectedFields?: ContentFieldId[]) => {
+      setSelectedContentTypes((prevState) => [...prevState, id]);
+      setSelectedContentFields((prevState) => {
+        prevState[id] = preSelectedFields ?? compatibility.compatibleFields[id];
+
+        return prevState;
+      });
+    },
+    [compatibility.compatibleFields]
+  );
+
+  const toggleContentType = (id: ContentTypeId) => {
+    if (selectedContentTypes.includes(id)) {
+      toggleContentTypeOff(id);
+    } else {
+      toggleContentTypeOn(id);
+    }
+  };
+
+  const toggleField = (contentTypeId: ContentTypeId, fieldId: ContentFieldId) => {
+    const currentlySelectedFields = selectedContentFields[contentTypeId];
+    const isContentTypeSelected = selectedContentTypes.includes(contentTypeId);
+    const isFieldSelected = currentlySelectedFields?.includes(fieldId);
+
+    if (isFieldSelected) {
+      if (currentlySelectedFields.length === 1) {
+        toggleContentTypeOff(contentTypeId);
+
+        return;
+      }
+
+      setSelectedContentFields({
+        ...selectedContentFields,
+        [contentTypeId]: currentlySelectedFields.filter((field) => field !== fieldId),
+      });
+    } else {
+      if (!isContentTypeSelected) {
+        toggleContentTypeOn(contentTypeId, [fieldId]);
+
+        return;
       }
+
+      setSelectedContentFields({
+        ...selectedContentFields,
+        [contentTypeId]: [...currentlySelectedFields, fieldId],
+      });
+    }
+  };
+
+  const selectMany = useCallback(
+    (ids: ContentTypeId[], preSelectedFields = compatibility.compatibleFields) => {
+      const fieldsToSelect = Object.fromEntries(Object.entries(preSelectedFields).filter(([id]) => ids.includes(id as ContentTypeId)));
+
+      setSelectedContentTypes(ids);
+      setSelectedContentFields(fieldsToSelect);
     },
-    [isSelected]
+    [compatibility]
   );
 
-  const SelectionComponent =
-    richTextFields.length > 1
-      ? () => <FieldSelection selectedFields={selectedFields} richTextFields={richTextFields} toggleSelection={toggleSelection} isSelected={isSelected} />
-      : () => null;
+  const selectAllCompatible = useCallback(() => {
```

Comment:
Some functions are wrapped in `useCallback`, and some are not. Is there a reason for this?

Reply:
Those that do are wrapped around the `compatibility` object passed into the hook. `useCallback` isn't absolutely necessary here imo, but then they're called in `useEffect` in a component and they're in the deps array, because linter complains otherwise... And of course without `useCallback` the `useEffect` runs over and over again. So basically I opted for this rather than `es-lint disable ...`.